smrt/inputs/make_bedrock.py

```diff
@@ -0,0 +1,236 @@
+    :param temperature: The physical temperature of the bedrock (K). Mandatory.
+    :param frequency_Hz: The frequency (Hz) to use when calculating complex permittivity for conductivity-based models. 
+        Note that this parameter is ignored by models with fixed measurement frequencies (e.g., "christianson"). Optional.
+    :param kwargs: Additional keyword arguments passed to the final substrate model constructor.
+    :raises SMRTError: If required parameters are missing or models are unrecognized.
+    :return: An instance of a Substrate model.
+    """
+
+    if temperature is None:
+        raise SMRTError("Temperature must be provided for bedrock.")
+
+    # 1. Select the bedrock
+    if isinstance(permittivity_model, str):
+        if permittivity_model in NAMED_PERMITTIVITY_MODELS:
+            model_name = permittivity_model
+            model_func = NAMED_PERMITTIVITY_MODELS[model_name]
+
+            if bedrock_type is None:
+                if model_name == 'hartlieb':
+                    valid_types = list(HARTLIEB_2016_DATA.keys())
+                elif model_name == 'christianson':
+                    valid_types = list(CHRISTIANSON_2016_DATA.keys())
+                else:
+                    valid_types = ['(refer to documentation)']
+
+                raise SMRTError(
+                    f"The material type for '{model_name}' must be provided via the 'bedrock_type' argument. "
+                    f"Valid types include: {', '.join(valid_types)}."
+                )
+
+            permittivity_value = model_func(bedrock_type)
+            permittivity_model = _get_constant_permittivity_model(permittivity_value)
+            
+        else:
+            raise SMRTError(f"The permittivity model name '{permittivity_model}' is not recognized. "
+                            f"Choose from: {', '.join(NAMED_PERMITTIVITY_MODELS.keys())}, a number, or a callable function.")
+            
+    # 2. Handle constant number (real or complex) permittivity for 
+    # same compatibility as make_soil
+    elif isinstance(permittivity_model, Number):  # a constant value
+        permittivity_model = _get_constant_permittivity_model(permittivity_model)
+        
+    # 3. Handle callable permittivity model (which is already a function)
+    # same as make soil
+    elif not callable(permittivity_model):
+        raise SMRTError("The permittivity_model argument is not of the accepted types: "
+                        "a string, a number (real/complex), or a callable function.")
+
+    # process the substrate_model argument (assumes get_substrate_model returns a class/type)
+    if not isinstance(substrate_model, type):
+        substrate_model = get_substrate_model(substrate_model)
+
+    # create the instance
+    return substrate_model(temperature, permittivity_model, **kwargs)
+
+
+# --- Private function ---
+
+def _get_constant_permittivity_model(constant_value: complex | Number) -> Callable[[float, float], complex | Number]:
```

Comment:
it's not working the way one usually think (see e.g. https://stackoverflow.com/questions/21053988/lambda-function-accessing-outside-variable ).
Anyway, this is not needed anymore, as I've implemented very recently this function in the SubstrateBase class directly, so it applies to any substrates.